### PR TITLE
Support for Flattening

### PR DESCRIPTION
**Pull request by mitsuhiko**
Third time is the charm. This replaces #1177 and #1178 and adds support for `#[serde(flatten)]`. This pull request can provide the features for #119 as well as #941.

## What it does

This pull request adds support for flattening:

* It introduces a `#[serde(flatten)]` field attribute which tells serde to flatten the container.
* It introduces an internal flag on a struct container that indicates that flattening is wanted on the container (at least one field has the flatten attribute).

Currently I only added support for maps and structs but I think this can be extended further to more types (including newtypes). Additionally right now the key must be a string. This requirement does not make a lot of sense and most likely any type of key can be supported with a bit of extra work.

Because of current limitations on the serializer interface structs that use flattening will be serialized as maps and not as structs.

## Example usage

```rust
use std::collections::HashMap;

#[derive(Serialize, Deserialize, Debug)]
pub struct Context {
    info: String,
    #[serde(flatten)]
    data: Data,
    #[serde(flatten)]
    extra: HashMap<String, String>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Data {
    foo: String,
    bar: String,
}
```

JSON:

```json
{
  "info": "foo bar baz",
  "foo": "foo value",
  "bar": "bar value",
  "extra_key": "extra_value"
}
```

Rust debug output of struct:

```rust
Context {
    info: "foo bar baz",
    data: Data {
        foo: "foo value",
        bar: "bar value"
    },
    extra: {
        "extra_key": "extra_value"
    }
}
```

## What's done

* [x] `#[serde(flatten)]` on map types
* [x] `#[serde(flatten)]` on structs
* [x] `#[serde(flatten)]` on enum structs (without special tagging)
* [x] tests
* [x] better error messages if flatten is used on incompatible types
* [x] `#[serde(flatten)]` on newtype
* [x] support for in-place deserialization `feature = "deserialize_in_place"`
* [x] non string key support

Thanks so much @dtolnay and @oli-obk for the help in getting this to the current state.

## Notes for the reviewer

* non string key support generates out a bunch of visit methods (`visit_i32` etc.) for the field. This seems wasteful. Is there a better way to do this? In case not, maybe it's okay to throw away the handling for anything other than string and bytes?
* Error messages are largely formatted with `format_args!` at the moment. Some of the current helper methods do not work with unknown field info.
* bytes/string borrowing for keys involved in flattening is currently not implemented
* in-place deserialization for flattening is not implemented
* some helpers for `Content` are now public internally so I can reuse them.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
There are two issues with the current implementation I'm not sure how to resolve. One is that you cannot `#[serde(flatten)]` something which in itself is also `#[serde(flatten)]`. The second one is that you cannot `#[serde(flatten)]` an enum that has `#[serde(tag = "some_field")]`.

The reason for both of them is the same: the inner type buffers up the values for the outer type as well. The solution I can see is that both generated deserializers would have to share the same buffer which would require a method on the `Deserializer` struct.

I'm not sure yet if that works but the idea is that instead of the `TaggedContentVisitor` and the `FlatMapDeserializer` both buffering there would have to be an abstraction in place. When a flatten level is reached or the `TaggedContentVisitor` would have to start buffering into a `Content` we would instead generate a wrapper type `X` that is either `X::Value(Content)` or `X::MapEntries(Vec<Option<(Content, Content)>>)` or something like that.

Then the generated enum deserializer would have to serializer over some sort of `TaggedSubsetContentDeserializer` which takes that wrapper type and the list of known fields and then only pull the known fields form it and leave the rest in the wrapper. For the non map case a `X::Value(Content)` is required.

Alternatively maybe `#[serde(tag = "some_field")]` should just be outright rejected in case of flatten for now?

**Comment by mitsuhiko** [comment]
I killed the `#[serde(repr = "map")]` thing. If someone wants this it can be brought back easily. For now it just makes things more noisy.

**Comment by mitsuhiko** [comment]
One more thing I think I want to do (does not have to be now) is to make the `__Field::__other(String)` into a `__Field::__other(Content<'de>)`.

File: serde/src/private/ser.rs [hunk 1]
```diff
@@ -58,10 +63,10 @@ enum Unsupported {
     ByteArray,
     Optional,
     Unit,
+    UnitStruct,
```
Comment (dtolnay) [inline serde/src/private/ser.rs]: I get a dead code warning in no-default-features mode. Can we cfg away this variant, or is this a bug?

```
$ cargo build --no-default-features

warning: variant is never constructed: `UnitStruct`
  --> serde/src/private/ser.rs:66:5
   |
66 |     UnitStruct,
   |     ^^^^^^^^^^
   |
   = note: #[warn(dead_code)] on by default
```

File: serde/src/private/ser.rs [hunk 2]
```diff
@@ -1028,3 +1033,282 @@ mod content {
...
+        self,
+        _: &'static str,
+        _: u32,
+        _: &'static str,
+        value: &T,
+    ) -> Result<Self::Ok, Self::Error>
+    where
+        T: Serialize,
+    {
+        value.serialize(self)
```
Comment (dtolnay) [inline serde/src/private/ser.rs]: I don't think this is what I would expect. The variant name is relevant.

```rust
#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_json;

use std::collections::BTreeMap;

#[derive(Serialize)]
struct S {
    #[serde(flatten)]
    flat: E,
}

#[derive(Serialize)]
enum E {
    Q(BTreeMap<String, String>),
}

fn main() {
    let e = E::Q({
        let mut map = BTreeMap::new();
        map.insert("k".to_owned(), "v".to_owned());
        map
    });
    println!("{}", serde_json::to_string(&e).unwrap());

    let s = S { flat: e };
    println!("{}", serde_json::to_string(&s).unwrap());
}
```

```json
{"Q":{"k":"v"}}
{"k":"v"}
```

Without flatten it is `{"flat":{"Q":{"k":"v"}}}` so with flatten I would expect `{"Q":{"k":"v"}}`.

File: serde_derive/src/ser.rs [hunk 3]
```diff
@@ -279,6 +287,44 @@ fn serialize_struct(params: &Parameters, fields: &[Field], cattrs: &attr::Contai
...
+
+    let mut serialized_fields = fields
+        .iter()
+        .filter(|&field| !field.attrs.skip_serializing())
+        .peekable();
+
+    let let_mut = mut_if(serialized_fields.peek().is_some());
+
+    let len = if cattrs.has_flatten() {
+        quote!(None)
```
Comment (dtolnay) [inline serde_derive/src/ser.rs]: This is really unfortunate but please make sure all paths in the generated code are fully qualified relative to `_serde`. Here it would be `_serde::export::None`. Otherwise we get weird failures depending on what the users have in scope (#737).

File: serde_derive/src/ser.rs [hunk 4]
```diff
@@ -877,20 +924,33 @@ fn serialize_struct_visitor(
             }
 
             let span = Span::def_site().located_at(field.original.span());
-            let func = struct_trait.serialize_field(span);
-            let ser = quote! {
-                try!(#func(&mut __serde_state, #key_expr, #field_expr));
+            let ser = if field.attrs.flatten() {
+                quote! {
+                    try!((#field_expr).serialize(_serde::private::ser::FlatMapSerializer(&mut __serde_state)));
```
Comment (dtolnay) [inline serde_derive/src/ser.rs]: Please use fully qualified [UFCS](https://doc.rust-lang.org/book/first-edition/ufcs.html) for the `serialize` method call. Otherwise this is screwed up by other traits the user may have in scope.

File: serde_derive/src/ser.rs [hunk 5]
```diff
@@ -877,20 +924,33 @@ fn serialize_struct_visitor(
             }
 
             let span = Span::def_site().located_at(field.original.span());
-            let func = struct_trait.serialize_field(span);
-            let ser = quote! {
-                try!(#func(&mut __serde_state, #key_expr, #field_expr));
+            let ser = if field.attrs.flatten() {
```
Comment (dtolnay) [inline serde_derive/src/ser.rs]: This fails to compile if we are inside a struct variant. Could you handle that case if it is easy, or produce an error instead?

```rust
#[derive(Serialize)]
enum E {
    Q {
        #[serde(flatten)]
        flat: BTreeMap<String, String>,
    },
}
```

```
error[E0277]: the trait bound `<__S as serde::Serializer>::SerializeStructVariant: serde::ser::SerializeMap` is not satisfied
 --> src/main.rs:9:10
  |
9 | #[derive(Serialize)]
  |          ^^^^^^^^^ the trait `serde::ser::SerializeMap` is not implemented for `<__S as serde::Serializer>::SerializeStructVariant`
  |
  = help: consider adding a `where <__S as serde::Serializer>::SerializeStructVariant: serde::ser::SerializeMap` bound
  = note: required because of the requirements on the impl of `serde::Serializer` for `serde::private::ser::FlatMapSerializer<'_, <__S as serde::Serializer>::SerializeStructVariant>`
```

**Comment by dtolnay** [comment]
I have reviewed everything except serde_derive/src/de.rs. Looking great! Will continue tomorrow.

File: serde_derive/src/de.rs [hunk 6]
```diff
@@ -2251,15 +2476,45 @@ fn deserialize_map_in_place(
...
+                let #name: #field_ty = try!(_serde::de::Deserialize::deserialize(
+                    _serde::private::de::FlatMapDeserializer(
+                        &mut __collect,
+                        _serde::export::PhantomData)));
+            }
+        });
+
+    let collected_deny_unknown_fields = if cattrs.has_flatten() && cattrs.deny_unknown_fields() {
+        Some(quote! {
+            if let Some(Some((__key, _))) = __collect.into_iter().filter(|x| x.is_some()).next() {
```
Comment (oli-obk) [inline serde_derive/src/de.rs]: do we need to use UFCs here, too?

File: serde_derive_internals/src/attr.rs [hunk 7]
```diff
@@ -878,6 +891,11 @@ impl Field {
                         }
                     }
 
+                    // Parse `#[serde(skip_deserializing)]`
```
Comment (oli-obk) [inline serde_derive_internals/src/attr.rs]: copy paste comment

File: test_suite/tests/test_annotations.rs [hunk 8]
```diff
@@ -1267,3 +1318,382 @@ fn test_from_into_traits() {
     assert_ser_tokens::<StructFromEnum>(&StructFromEnum(None), &[Token::None]);
     assert_de_tokens::<StructFromEnum>(&StructFromEnum(Some(2)), &[Token::Some, Token::U32(2)]);
 }
+
+#[test]
+fn test_collect_other() {
+    let mut extra = HashMap::new();
+    extra.insert("c".into(), 3);
+    assert_de_tokens(
```
Comment (oli-obk) [inline test_suite/tests/test_annotations.rs]: use assert_tokens so you don't have to repeat the `ser` fn call below

File: test_suite/tests/test_annotations.rs [hunk 9]
```diff
@@ -1267,3 +1318,382 @@ fn test_from_into_traits() {
...
+        d: String,
+        e: Option<u64>,
+    }
+
+    #[derive(Debug, PartialEq, Serialize, Deserialize)]
+    struct Second {
+        f: u32,
+    }
+
+    assert_de_tokens(
```
Comment (oli-obk) [inline test_suite/tests/test_annotations.rs]: same here, `assert_tokens`

Comment (mitsuhiko) [reply test_suite/tests/test_annotations.rs]: can't do on this one because of the involved `Option`. The code above I will change. That's a leftover from refactoring when an `Option` was used there as well.

File: serde_derive/src/de.rs [hunk 10]
```diff
@@ -268,7 +268,11 @@ fn deserialize_in_place_body(cont: &Container, params: &Parameters) -> Option<St
 
     let code = match cont.data {
         Data::Struct(Style::Struct, ref fields) => {
-            deserialize_struct_in_place(None, params, fields, &cont.attrs, None)
+            if let Some(code) = deserialize_struct_in_place(None, params, fields, &cont.attrs, None) {
```
[the remark this answers is not in the document]
Comment (mitsuhiko) [reply serde_derive/src/de.rs]: Yep, but as far as I can tell serde does not use `?` in the codebase.

File: serde/src/private/de.rs [hunk 11]
```diff
@@ -2197,14 +2199,18 @@ impl<'a, 'de, E> MapAccess<'de> for FlatMapAccess<'a, 'de, E>
         T: DeserializeSeed<'de>,
     {
         while let Some(item) = self.iter.next() {
```
Comment (oli-obk) [inline serde/src/private/de.rs]: I still find this logic hard to follow, can you at least add some comments so future contributors know what's going on

File: serde_derive/src/de.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (dtolnay) [reply serde_derive/src/de.rs]: Yeah it showed up noticeably in benchmarks last time I tried it -- https://github.com/rust-lang/rust/issues/37939. It would be fine to use `?` in serde_derive but not in serde or the generated code, but it has just been simpler to use `try!` everywhere until the performance issue is figured out. That way we never have to argue about whether a particular use of `?` ever matters for performance.

**Comment by mitsuhiko** [comment]
I think I acted on all comments.

**Comment by Ralith** [comment]
Was this tested on tuple structs? Flattening newtypes specifically is the overwhelmingly significant usecase for me here.

**Comment by dtolnay** [comment]
@Ralith please file an issue for flattening newtypes.

**Comment by theduke** [comment]
@mitsuhiko really wanted this as well, thanks for pushing it through.
